pkg/controllers/management/k3supgrade/deployPlans.go

```diff
@@ -0,0 +1,166 @@
+package k3supgrade
+
+import (
+	"reflect"
+
+	planv1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	planClientset "github.com/rancher/system-upgrade-controller/pkg/generated/clientset/versioned/typed/upgrade.cattle.io/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// deployPlans creates a master and worker plan in the downstream cluster to instrument
+// the system-upgrade-controller in the downstream cluster
+func (h *handler) deployPlans(cluster *v3.Cluster) error {
+
+	// access downstream cluster
+	clusterCtx, err := h.manager.UserContext(cluster.Name)
+	if err != nil {
+		return err
+
+	}
+
+	// create a client for GETing Plans in the downstream cluster
+	planConfig, err := planClientset.NewForConfig(&clusterCtx.RESTConfig)
+	if err != nil {
+		return err
+	}
+	planClient := planConfig.Plans(metav1.NamespaceAll)
+
+	planList, err := planClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	masterPlan := planv1.Plan{}
+	workerPlan := planv1.Plan{}
+	// deactivate all existing plans that are not managed by Rancher
+	for _, plan := range planList.Items {
+		if _, ok := plan.Labels[rancherManagedPlan]; !ok {
+			// inverse selection is used here, we select a non-existent label
+			plan.Spec.NodeSelector = &metav1.LabelSelector{
```

Comment:
I feel like we should append this to the existing match expressions so as to be less destructive.

Reply:
agree

Reply:
Resolved